src/pstore/pstore.c

```diff
@@ -0,0 +1,399 @@
+                 * with the underlying pstore backend (ie. EFI may be ~2KiB),
+                 * which means an example pstore with approximately 64KB of
+                 * storage may have up to roughly 32 dmesg files that could be
+                 * related, depending upon the size of the original dmesg.
+                 * Here we look at the dmesg filename and try to discern if
+                 * files are part of a related group, meaning the same original
+                 * dmesg.
+                 * The two known pstore backends are EFI and ERST. These backends
+                 * store data in the Common Platform Error Record, CPER, format.
+                 * The dmesg- filename contains the CPER record id, a 64bit number
+                 * (in decimal notation). In Linux, the record id is encoded with
+                 * two digits for the dmesg part (chunk) number and 3 digits for
+                 * the count number. So allowing an additional digit to compensate
+                 * for advancing time, this code ignores the last six digits of the
+                 * filename in determining the record id.
+                 * For the EFI backend, the record id encodes an id
+                 * in the upper 32 bits, and a timestamp in the lower 32-bits.
+                 * So ignoring the least significant 6 digits has proven to
+                 * generally identify related dmesg entries.  */
+#define PSTORE_FILENAME_IGNORE 6
+
+                /* extract common portion of record id */
+                current_id = NULL;
+                if (strlen(p) > PSTORE_FILENAME_IGNORE) {
+                        current_id = (id == id1) ? id2 : id1;
+                        strcpy(current_id, p+1);
+                        current_id[strlen(current_id) - PSTORE_FILENAME_IGNORE] = '\0';
+                }
+
+                /* Now move file from pstore to archive storage */
+                move_file(pe, current_id);
+
+                /* If the current record id is NOT the same as the
+                 * previous record id, then start a new dmesg.txt file */
+                if (!streq(current_id, id)) {
+                        /* Encountered a new dmesg group, close out old one, open new one */
+                        if (dmesg) {
+                                (void) write_dmesg(dmesg, dmesg_size, id);
+                                free(dmesg);
+                                dmesg = NULL;
+                                dmesg_size = 0;
+                        }
+
+                        /* Swap record id pointers */
+                        id[0] = '\0'; /* invalidate previous record id */
+                        id = (id == id1) ? id2 : id1; /* make current id */
+                }
+
+                /* Reconstruction of dmesg is done as a useful courtesy, do not log errors */
+                dmesg2 = realloc(dmesg, dmesg_size + strlen(pe->dirent.d_name) + strlen(":\n") + 1);
```

Comment:
why do two realloc()s here, instead of one?

Reply:
done